byway/src/main/java/com/google/sps/servlets/DestinationsServlet.java

```diff
@@ -0,0 +1,87 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.common.flogger.FluentLogger;
+
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.ArrayList;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.annotation.MultipartConfig;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+/** Servlet that returns start location and destinations user inputs */
+@MultipartConfig
+@WebServlet("/api/destinations")
+public class DestinationsServlet extends HttpServlet {
+
+  private final Gson gson = new Gson();
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
+  private Key userKey;
+
+  @Override
+  public void init() {
+    Entity userEntity = new Entity("UserInputs");
+    userEntity.setProperty("start", "");
+    userEntity.setProperty("destinations", new ArrayList<String>());
+    datastore.put(userEntity);
+    userKey= userEntity.getKey();
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Entity entity; 
+    try{ 
+      entity = datastore.get(userKey);
```

Comment:
Try to limit the scope of catch blocks to only cover the line where an error is expected to occur. In this case, you can only get EntityNotFound on the call to `datastore.get`, so there's no need to have the rest of the logic for the method inside the try block.

```java
Entity entity;
try {
  entity = datastore.get(userKey);
} catch (...) {
  ...
}
// operations on the entity.
```

We do this to avoid handling errors that we didn't intend to handle. Catching too many errors can cause confusion and make issues harder to debug. Consider what would happen if some other code inside this method tried to access something from datastore and had an error, e.g. if you had another calss with some shared code which happened to also access datastore. You could catch errors from that code, but the log message will always say ""Unable to find UserLocations Entity" which would be confusing to the person trying to debug it.

Reply:
fixed

Reply:
Note that if there's no sensible thing to do in the case that the entity is missing you can just `return` from the `catch` block.

Keep in mind that if you print no response during the execution of the doGet method, nothing will be sent to the javascript fetch. Note that that's different from sending an empty `UserLocations`; sending no response means that there's nothing for `.then(response => response.json())` to convert from json.

One simple way you could handle this in the case of entity not found is to just print an empty `UserLocations`, so the JS side displays the default value.

Reply:
ok I had it return an empty UserLocations

Reply:
If you include an empty return statement at the end of the catch block, you don't have to worry about whether `entity` is null below there, since you know that it won't be null. In fact, if you do this, you can leave out the `entity = null;` default value, because the compiler can *prove* that the entity will have a value if the control flow of the program reaches the code below the catch block.

```java
    } catch(EntityNotFoundException e){
      UserLocations userLocations = new UserLocations("", new ArrayList<String>()); 
      response.setContentType("application/json;");
      response.getWriter().println(gson.toJson(userLocations)); 
      logger.atInfo().withCause(e).log("Unable to find UserLocations Entity %s", userKey);
      return; // This means that if `entity` is not found, you return from the function without executing any code below, and if you execute any code below the catch block, `entity` must not be null (so you don't need to null check it).
    }
```

Reply:
ah makes sense, fixed it!